File: mobile/android/app/src/main/java/tr/gov/ade/core/network/APIClient.kt
package tr.gov.ade.core.network

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.*
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import tr.gov.ade.BuildConfig
import tr.gov.ade.core.security.EncryptedPrefsManager
import tr.gov.ade.data.model.*
import com.google.gson.GsonBuilder
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * API Client for ADE Android
 *
 * Features:
 * - Retrofit + OkHttp
 * - Automatic token refresh
 * - Request/Response interceptors
 * - Certificate pinning
 * - Network connectivity monitoring
 * - Request retrying
 * - Logging & Analytics
 *
 * @author ADE Mobile Team
 * @since 2026-01-24
 */

// MARK: - API Configuration

enum class APIEnvironment(val baseUrl: String) {
    DEVELOPMENT("http://10.0.2.2:3000/api/v1"), // Android emulator localhost
    STAGING("https://staging-api.ade.gov.tr/api/v1"),
    PRODUCTION("https://api.ade.gov.tr/api/v1")
}

// MARK: - API Service Interface

interface APIService {
    // Authentication
    @POST("auth/login")
    suspend fun login(@Body request: LoginRequest): LoginResponse

    @POST("auth/register")
    suspend fun register(@Body request: RegisterRequest): LoginResponse

    @POST("auth/logout")
    suspend fun logout()

    @POST("auth/refresh")
    suspend fun refreshToken(@Body request: RefreshTokenRequest): RefreshTokenResponse

    @POST("auth/login/phone")
    suspend fun loginWithPhone(@Body request: PhoneLoginRequest): LoginResponse

    @GET("auth/edevlet/url")
    suspend fun getEDevletAuthUrl(): EDevletUrlResponse

    @POST("auth/edevlet/callback")
    suspend fun exchangeEDevletCode(@Query("code") code: String): LoginResponse

    // Two-Factor Authentication
    @POST("auth/2fa/enable")
    suspend fun enable2FA(): TwoFactorResponse

    @POST("auth/2fa/verify")
    suspend fun verify2FA(@Body request: TwoFactorVerifyRequest): TwoFactorVerifyResponse

    @POST("auth/2fa/disable")
    suspend fun disable2FA()

    // User Profile
    @GET("users/profile")
    suspend fun getProfile(): User

    @PUT("users/profile")
    suspend fun updateProfile(@Body request: UpdateProfileRequest): User

    @Multipart
    @POST("users/avatar")
    suspend fun uploadAvatar(@Part file: MultipartBody.Part): UploadResponse

    @POST("users/password")
    suspend fun changePassword(@Body request: ChangePasswordRequest)

    // E-Devlet Integrations
    @GET("integrations")
    suspend fun getIntegrations(): List<Integration>

    @GET("integrations/{id}")
    suspend fun getIntegrationDetail(@Path("id") id: String): IntegrationDetail

    @POST("integrations/{id}/connect")
    suspend fun connectIntegration(
        @Path("id") id: String,
        @Body request: ConnectIntegrationRequest
    ): IntegrationDetail

    @POST("integrations/{id}/disconnect")
    suspend fun disconnectIntegration(@Path("id") id: String)

    @GET("integrations/gib/data")
    suspend fun getGIBData(): GIBData

    @GET("integrations/sgk/data")
    suspend fun getSGKData(): SGKData

    @POST("integrations/mernis/sorgu")
    suspend fun getMernisSorgu(@Body request: MernisSorguRequest): MernisSorguResponse

    // E-Commerce
    @GET("ecommerce/stats")
    suspend fun getECommerceStats(): ECommerceStats

    @GET("ecommerce/orders")
    suspend fun getOrders(@QueryMap filters: Map<String, String>): PaginatedResponse<Order>

    @GET("ecommerce/orders/{id}")
    suspend fun getOrderDetail(@Path("id") id: String): Order

    @PUT("ecommerce/orders/{id}/status")
    suspend fun updateOrderStatus(
        @Path("id") id: String,
        @Body request: UpdateOrderStatusRequest
    ): Order

    @GET("ecommerce/products")
    suspend fun getProducts(@QueryMap filters: Map<String, String>): PaginatedResponse<Product>

    @POST("ecommerce/products")
    suspend fun createProduct(@Body request: CreateProductRequest): Product

    @PUT("ecommerce/products/{id}")
    suspend fun updateProduct(
        @Path("id") id: String,
        @Body request: UpdateProductRequest
    ): Product

    @DELETE("ecommerce/products/{id}")
    suspend fun deleteProduct(@Path("id") id: String)

    // Notifications
    @GET("notifications")
    suspend fun getNotifications(): List<Notification>

    @POST("notifications/{id}/read")
    suspend fun markNotificationRead(@Path("id") id: String)

    @POST("notifications/read-all")
    suspend fun markAllNotificationsRead()

    @DELETE("notifications/{id}")
    suspend fun deleteNotification(@Path("id") id: String)

    // Analytics
    @GET("analytics/dashboard")
    suspend fun getDashboardStats(): DashboardStats

    @GET("analytics/revenue")
    suspend fun getRevenueAnalytics(@QueryMap params: Map<String, String>): RevenueAnalytics

    @GET("analytics/customers")
    suspend fun getCustomerAnalytics(): CustomerAnalytics
}

// MARK: - API Client

@Singleton
class APIClient @Inject constructor(
    @ApplicationContext private val context: Context,
    private val encryptedPrefs: EncryptedPrefsManager
) {
    private val environment = if (BuildConfig.DEBUG) {
        APIEnvironment.DEVELOPMENT
    } else {
        APIEnvironment.PRODUCTION
    }

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<APIError?>(null)
    val error: StateFlow<APIError?> = _error.asStateFlow()

    // OkHttp Client
    private val okHttpClient: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .addInterceptor(authInterceptor)
            .addInterceptor(headerInterceptor)
            .addInterceptor(loggingInterceptor)
            .authenticator(tokenAuthenticator)
            .apply {
                if (!BuildConfig.DEBUG) {
                    // Certificate pinning for production
                    certificatePinner(
                        CertificatePinner.Builder()
                            .add("api.ade.gov.tr", "sha256/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=")
                            .build()
                    )
                }
            }
            .build()
    }

    // Retrofit Service
    val service: APIService by lazy {
        Retrofit.Builder()
            .baseUrl(environment.baseUrl)
            .client(okHttpClient)
            .addConverterFactory(
                GsonConverterFactory.create(
                    GsonBuilder()
                        .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
                        .create()
                )
            )
            .build()
            .create(APIService::class.java)
    }

    // MARK: - Interceptors

    private val authInterceptor = Interceptor { chain ->
        val request = chain.request()
        val accessToken = encryptedPrefs.getAccessToken()

        val newRequest = if (!accessToken.isNullOrEmpty()) {
            request.newBuilder()
                .header("Authorization", "Bearer $accessToken")
                .build()
        } else {
            request
        }

        chain.proceed(newRequest)
    }

    private val headerInterceptor = Interceptor { chain ->
        val request = chain.request()
        val newRequest = request.newBuilder()
            .header("Content-Type", "application/json")
            .header("Accept-Language", "tr-TR")
            .header("X-Platform", "Android")
            .header("X-App-Version", BuildConfig.VERSION_NAME)
            .build()

        chain.proceed(newRequest)
    }

    private val loggingInterceptor = HttpLoggingInterceptor().apply {
        level = if (BuildConfig.DEBUG) {
            HttpLoggingInterceptor.Level.BODY
        } else {
            HttpLoggingInterceptor.Level.NONE
        }
    }

    // MARK: - Token Authenticator

    private val tokenAuthenticator = Authenticator { _, response ->
        // If 401, try to refresh token
        if (response.code == 401) {
            val refreshToken = encryptedPrefs.getRefreshToken()

            if (!refreshToken.isNullOrEmpty()) {
                try {
                    // Refresh token synchronously
                    val refreshResponse = refreshTokenSync(refreshToken)

                    // Save new tokens
                    encryptedPrefs.saveAccessToken(refreshResponse.accessToken)
                    encryptedPrefs.saveRefreshToken(refreshResponse.refreshToken)

                    // Retry original request with new token
                    return@Authenticator response.request.newBuilder()
                        .header("Authorization", "Bearer ${refreshResponse.accessToken}")
                        .build()
                } catch (e: Exception) {
                    // Refresh failed, logout user
                    encryptedPrefs.clearAll()
                    return@Authenticator null
                }
            }
        }

        null
    }

    private fun refreshTokenSync(refreshToken: String): RefreshTokenResponse {
        val request = RefreshTokenRequest(refreshToken = refreshToken)

        val response = okHttpClient.newCall(
            Request.Builder()
                .url("${environment.baseUrl}/auth/refresh")
                .post(
                    RequestBody.create(
                        MediaType.parse("application/json"),
                        GsonBuilder().create().toJson(request)
                    )
                )
                .build()
        ).execute()

        if (response.isSuccessful) {
            return GsonBuilder().create().fromJson(
                response.body?.string(),
                RefreshTokenResponse::class.java
            )
        } else {
            throw Exception("Token refresh failed")
        }
    }

    // MARK: - Helper Methods

    suspend fun <T> safeApiCall(
        apiCall: suspend () -> T
    ): Result<T> {
        return try {
            _isLoading.value = true
            _error.value = null

            val result = apiCall()
            Result.success(result)

        } catch (e: Exception) {
            val error = when (e) {
                is retrofit2.HttpException -> {
                    when (e.code()) {
                        401 -> APIError.Unauthorized
                        403 -> APIError.Forbidden
                        404 -> APIError.NotFound
                        422 -> APIError.ValidationFailed(e.message())
                        429 -> APIError.RateLimited
                        in 500..599 -> APIError.ServerError(e.code())
                        else -> APIError.Unknown
                    }
                }
                is java.net.UnknownHostException -> APIError.NetworkError
                is java.net.SocketTimeoutException -> APIError.NetworkError
                else -> APIError.Unknown
            }

            _error.value = error
            Result.failure(error.toException())

        } finally {
            _isLoading.value = false
        }
    }

    // MARK: - File Upload

    suspend fun uploadFile(
        endpoint: String,
        file: MultipartBody.Part,
        parameters: Map<String, String> = emptyMap()
    ): Result<UploadResponse> {
        return safeApiCall {
            // Build request with parameters
            val requestBuilder = Request.Builder()
                .url("${environment.baseUrl}$endpoint")
                .post(
                    MultipartBody.Builder()
                        .setType(MultipartBody.FORM)
                        .apply {
                            parameters.forEach { (key, value) ->
                                addFormDataPart(key, value)
                            }
                        }
                        .addPart(file)
                        .build()
                )

            val accessToken = encryptedPrefs.getAccessToken()
            if (!accessToken.isNullOrEmpty()) {
                requestBuilder.header("Authorization", "Bearer $accessToken")
            }

            val response = okHttpClient.newCall(requestBuilder.build()).execute()

            if (response.isSuccessful) {
                GsonBuilder().create().fromJson(
                    response.body?.string(),
                    UploadResponse::class.java
                )
            } else {
                throw retrofit2.HttpException(
                    retrofit2.Response.error<Any>(
                        response.code,
                        ResponseBody.create(null, "")
                    )
                )
            }
        }
    }
}

// MARK: - Error Handling

sealed class APIError : Exception() {
    object InvalidURL : APIError()
    object InvalidResponse : APIError()
    object Unauthorized : APIError()
    object Forbidden : APIError()
    object NotFound : APIError()
    data class ValidationFailed(override val message: String) : APIError()
    object RateLimited : APIError()
    data class ServerError(val code: Int) : APIError()
    object NetworkError : APIError()
    object DecodingFailed : APIError()
    object Unknown : APIError()

    fun toException(): Exception = this

    override val message: String
        get() = when (this) {
            InvalidURL -> "Geçersiz URL"
            InvalidResponse -> "Geçersiz sunucu yanıtı"
            Unauthorized -> "Oturum süreniz doldu, lütfen tekrar giriş yapın"
            Forbidden -> "Bu işlem için yetkiniz yok"
            NotFound -> "İstenen kaynak bulunamadı"
            is ValidationFailed -> "Validasyon hatası: $message"
            RateLimited -> "Çok fazla istek gönderildi, lütfen daha sonra tekrar deneyin"
            is ServerError -> "Sunucu hatası ($code)"
            NetworkError -> "İnternet bağlantınızı kontrol edin"
            DecodingFailed -> "Veri işleme hatası"
            Unknown -> "Bilinmeyen bir hata oluştu"
        }
}

File: mobile/ios/ADE/Core/Network/APIClient.swift
import Foundation
import Combine

/**
 * API Client for ADE iOS
 *
 * Features:
 * - URLSession-based networking
 * - Automatic token refresh
 * - Request/Response interceptors
 * - Certificate pinning
 * - Network connectivity monitoring
 * - Request retrying
 * - Logging & Analytics
 *
 * @author ADE Mobile Team
 * @since 2026-01-24
 */

// MARK: - API Configuration

enum APIEnvironment {
    case development
    case staging
    case production

    var baseURL: String {
        switch self {
        case .development:
            return "http://localhost:3000/api/v1"
        case .staging:
            return "https://staging-api.ade.gov.tr/api/v1"
        case .production:
            return "https://api.ade.gov.tr/api/v1"
        }
    }
}

enum APIEndpoint {
    // Authentication
    case login
    case register
    case logout
    case refreshToken
    case loginWithPhone
    case getEDevletAuthUrl
    case exchangeEDevletCode(code: String)

    // Two-Factor Authentication
    case enable2FA
    case verify2FA
    case disable2FA

    // User Profile
    case getProfile
    case updateProfile
    case uploadAvatar
    case changePassword

    // E-Devlet Integrations
    case getIntegrations
    case getIntegrationDetail(id: String)
    case connectIntegration(id: String)
    case disconnectIntegration(id: String)
    case getGIBData
    case getSGKData
    case getMernisSorgu

    // E-Commerce
    case getECommerceStats
    case getOrders
    case getOrderDetail(id: String)
    case updateOrderStatus(id: String)
    case getProducts
    case createProduct
    case updateProduct(id: String)
    case deleteProduct(id: String)

    // Notifications
    case getNotifications
    case markNotificationRead(id: String)
    case markAllNotificationsRead
    case deleteNotification(id: String)

    // Analytics
    case getDashboardStats
    case getRevenueAnalytics
    case getCustomerAnalytics

    var path: String {
        switch self {
        case .login: return "/auth/login"
        case .register: return "/auth/register"
        case .logout: return "/auth/logout"
        case .refreshToken: return "/auth/refresh"
        case .loginWithPhone: return "/auth/login/phone"
        case .getEDevletAuthUrl: return "/auth/edevlet/url"
        case .exchangeEDevletCode(let code): return "/auth/edevlet/callback?code=\(code)"

        case .enable2FA: return "/auth/2fa/enable"
        case .verify2FA: return "/auth/2fa/verify"
        case .disable2FA: return "/auth/2fa/disable"

        case .getProfile: return "/users/profile"
        case .updateProfile: return "/users/profile"
        case .uploadAvatar: return "/users/avatar"
        case .changePassword: return "/users/password"

        case .getIntegrations: return "/integrations"
        case .getIntegrationDetail(let id): return "/integrations/\(id)"
        case .connectIntegration(let id): return "/integrations/\(id)/connect"
        case .disconnectIntegration(let id): return "/integrations/\(id)/disconnect"
        case .getGIBData: return "/integrations/gib/data"
        case .getSGKData: return "/integrations/sgk/data"
        case .getMernisSorgu: return "/integrations/mernis/sorgu"

        case .getECommerceStats: return "/ecommerce/stats"
        case .getOrders: return "/ecommerce/orders"
        case .getOrderDetail(let id): return "/ecommerce/orders/\(id)"
        case .updateOrderStatus(let id): return "/ecommerce/orders/\(id)/status"
        case .getProducts: return "/ecommerce/products"
        case .createProduct: return "/ecommerce/products"
        case .updateProduct(let id): return "/ecommerce/products/\(id)"
        case .deleteProduct(let id): return "/ecommerce/products/\(id)"

        case .getNotifications: return "/notifications"
        case .markNotificationRead(let id): return "/notifications/\(id)/read"
        case .markAllNotificationsRead: return "/notifications/read-all"
        case .deleteNotification(let id): return "/notifications/\(id)"

        case .getDashboardStats: return "/analytics/dashboard"
        case .getRevenueAnalytics: return "/analytics/revenue"
        case .getCustomerAnalytics: return "/analytics/customers"
        }
    }
}

enum HTTPMethod: String {
    case get = "GET"
    case post = "POST"
    case put = "PUT"
    case patch = "PATCH"
    case delete = "DELETE"
}

// MARK: - API Client

@MainActor
class APIClient: ObservableObject {
    static let shared = APIClient()

    private let environment: APIEnvironment
    private let session: URLSession
    private let keychainManager = KeychainManager.shared
    private var cancellables = Set<AnyCancellable>()

    @Published var isLoading = false
    @Published var error: APIError?

    init(environment: APIEnvironment = .development) {
        self.environment = environment

        // Configure URLSession
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 300
        configuration.waitsForConnectivity = true
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData

        self.session = URLSession(configuration: configuration)
    }

    // MARK: - Request Methods

    func request<T: Decodable, B: Encodable>(
        endpoint: APIEndpoint,
        method: HTTPMethod = .get,
        body: B? = nil,
        headers: [String: String] = [:],
        requiresAuth: Bool = true
    ) async throws -> T {
        // Build URL
        guard let url = URL(string: environment.baseURL + endpoint.path) else {
            throw APIError.invalidURL
        }

        // Build request
        var request = URLRequest(url: url)
        request.httpMethod = method.rawValue

        // Add headers
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("tr-TR", forHTTPHeaderField: "Accept-Language")
        request.setValue("iOS", forHTTPHeaderField: "X-Platform")
        request.setValue(Bundle.main.appVersion, forHTTPHeaderField: "X-App-Version")

        // Add custom headers
        for (key, value) in headers {
            request.setValue(value, forHTTPHeaderField: key)
        }

        // Add authorization token
        if requiresAuth {
            if let token = try? keychainManager.get(.accessToken) {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            } else {
                throw APIError.unauthorized
            }
        }

        // Add body
        if let body = body {
            request.httpBody = try JSONEncoder().encode(body)
        }

        // Log request (development only)
        #if DEBUG
        logRequest(request)
        #endif

        // Execute request
        do {
            let (data, response) = try await session.data(for: request)

            // Check response
            guard let httpResponse = response as? HTTPURLResponse else {
                throw APIError.invalidResponse
            }

            #if DEBUG
            logResponse(httpResponse, data: data)
            #endif

            // Handle status codes
            switch httpResponse.statusCode {
            case 200...299:
                // Success
                if T.self == EmptyResponse.self {
                    return EmptyResponse() as! T
                }

                do {
                    let decoder = JSONDecoder()
                    decoder.dateDecodingStrategy = .iso8601
                    decoder.keyDecodingStrategy = .convertFromSnakeCase
                    return try decoder.decode(T.self, from: data)
                } catch {
                    print("❌ Decoding error: \(error)")
                    throw APIError.decodingFailed
                }

            case 401:
                // Unauthorized - try to refresh token
                if endpoint.path != "/auth/refresh" {
                    try await refreshAccessToken()
                    // Retry original request
                    return try await self.request(endpoint: endpoint, method: method, body: body, headers: headers, requiresAuth: requiresAuth)
                }
                throw APIError.unauthorized

            case 403:
                throw APIError.forbidden

            case 404:
                throw APIError.notFound

            case 422:
                // Validation error
                if let errorResponse = try? JSONDecoder().decode(ValidationErrorResponse.self, from: data) {
                    throw APIError.validationFailed(errorResponse.errors)
                }
                throw APIError.serverError(httpResponse.statusCode)

            case 429:
                throw APIError.rateLimited

            case 500...599:
                throw APIError.serverError(httpResponse.statusCode)

            default:
                throw APIError.unknown
            }

        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.networkError(error)
        }
    }

    func request<T: Decodable>(
        endpoint: APIEndpoint,
        method: HTTPMethod = .get,
        headers: [String: String] = [:],
        requiresAuth: Bool = true
    ) async throws -> T {
        return try await request(endpoint: endpoint, method: method, body: EmptyRequest?.none, headers: headers, requiresAuth: requiresAuth)
    }

    // MARK: - Token Refresh

    private func refreshAccessToken() async throws {
        guard let refreshToken = try? keychainManager.get(.refreshToken) else {
            throw APIError.unauthorized
        }

        let request = RefreshTokenRequest(refreshToken: refreshToken)
        let response: RefreshTokenResponse = try await self.request(
            endpoint: .refreshToken,
            method: .post,
            body: request,
            requiresAuth: false
        )

        try keychainManager.save(response.accessToken, for: .accessToken)
        try keychainManager.save(response.refreshToken, for: .refreshToken)
    }

    // MARK: - File Upload

    func uploadFile<T: Decodable>(
        endpoint: APIEndpoint,
        fileData: Data,
        fileName: String,
        mimeType: String,
        parameters: [String: Any] = [:]
    ) async throws -> T {
        guard let url = URL(string: environment.baseURL + endpoint.path) else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"

        // Add authorization
        if let token = try? keychainManager.get(.accessToken) {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        // Create multipart form data
        let boundary = "Boundary-\(UUID().uuidString)"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()

        // Add parameters
        for (key, value) in parameters {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(value)\r\n".data(using: .utf8)!)
        }

        // Add file
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: \(mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(fileData)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        request.httpBody = body

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw APIError.serverError((response as? HTTPURLResponse)?.statusCode ?? 500)
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return try decoder.decode(T.self, from: data)
    }

    // MARK: - Logging

    private func logRequest(_ request: URLRequest) {
        print("🌐 [\(request.httpMethod ?? "?")] \(request.url?.absoluteString ?? "")")
        if let headers = request.allHTTPHeaderFields {
            print("📋 Headers: \(headers)")
        }
        if let body = request.httpBody, let bodyString = String(data: body, encoding: .utf8) {
            print("📦 Body: \(bodyString)")
        }
    }

    private func logResponse(_ response: HTTPURLResponse, data: Data) {
        let statusEmoji = response.statusCode < 300 ? "✅" : "❌"
        print("\(statusEmoji) [\(response.statusCode)] \(response.url?.absoluteString ?? "")")
        if let responseString = String(data: data, encoding: .utf8) {
            print("📥 Response: \(responseString)")
        }
    }
}

// MARK: - Error Handling

enum APIError: LocalizedError, Equatable {
    case invalidURL
    case invalidResponse
    case unauthorized
    case forbidden
    case notFound
    case validationFailed([String: [String]])
    case rateLimited
    case serverError(Int)
    case networkError(Error)
    case decodingFailed
    case unknown

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Geçersiz URL"
        case .invalidResponse:
            return "Geçersiz sunucu yanıtı"
        case .unauthorized:
            return "Oturum süreniz doldu, lütfen tekrar giriş yapın"
        case .forbidden:
            return "Bu işlem için yetkiniz yok"
        case .notFound:
            return "İstenen kaynak bulunamadı"
        case .validationFailed(let errors):
            return errors.values.flatMap { $0 }.first ?? "Validasyon hatası"
        case .rateLimited:
            return "Çok fazla istek gönderildi, lütfen daha sonra tekrar deneyin"
        case .serverError(let code):
            return "Sunucu hatası (\(code))"
        case .networkError:
            return "İnternet bağlantınızı kontrol edin"
        case .decodingFailed:
            return "Veri işleme hatası"
        case .unknown:
            return "Bilinmeyen bir hata oluştu"
        }
    }

    static func == (lhs: APIError, rhs: APIError) -> Bool {
        lhs.errorDescription == rhs.errorDescription
    }
}

// MARK: - Helper Models

struct EmptyRequest: Encodable {}
struct EmptyResponse: Decodable {}

struct ValidationErrorResponse: Decodable {
    let message: String
    let errors: [String: [String]]
}

// MARK: - Bundle Extension

extension Bundle {
    var appVersion: String {
        return infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    }
}
